### PR TITLE
super_parameter support for `type_init_formals`

### PR DESCRIPTION
**Pull request by pq**
See: #3131

/cc @bwilkerson @scheglov

### CODE REVIEW

File: lib/src/rules/type_init_formals.dart [hunk 1]
```diff
@@ -47,29 +48,46 @@ class TypeInitFormals extends LintRule {
...
-        if (field != null) {
-          if (nodeType.type == field.type) {
-            rule.reportLint(nodeType);
-          }
-        }
+    if (nodeType == null) return;
+    var cls = node.thisOrAncestorOfType<ClassDeclaration>()?.declaredElement;
+    if (cls == null) return;
+
+    var field = cls.getField(node.identifier.name);
```
Comment (scheglov) [inline lib/src/rules/type_init_formals.dart:70]: Check `FieldFormalParameterElement.field` instead.

File: lib/src/rules/type_init_formals.dart [hunk 2]
```diff
@@ -47,29 +48,46 @@ class TypeInitFormals extends LintRule {
...
+
+  @override
+  void visitSuperFormalParameter(SuperFormalParameter node) {
+    var nodeType = node.type;
+    if (nodeType == null) return;
+    var cls = node.thisOrAncestorOfType<ClassDeclaration>()?.declaredElement;
+    if (cls == null) return;
+
+    var inheritedElement = context.inheritanceManager
+        .getInherited2(cls, Name(null, node.identifier.name));
```
Comment (scheglov) [inline lib/src/rules/type_init_formals.dart:87]: For super-formal parameters we don't care about field, the super-constructor can do whatever it want with these parameters. But we inherit the type from the parameter of the super-constructor, see `SuperFormalParameterElement.superConstructorParameter`

File: lib/src/rules/type_init_formals.dart (hunk 1 above)
Comment (pq) [reply lib/src/rules/type_init_formals.dart]: Nice.  Thanks!

File: lib/src/rules/type_init_formals.dart (hunk 2 above)
Comment (pq) [reply lib/src/rules/type_init_formals.dart]: Ah.  That makes sense.  Updated.  PTAL. Thanks!

**Review by pq — COMMENTED** [review]
Thanks!

File: lib/src/rules/type_init_formals.dart [hunk 3]
```diff
@@ -60,17 +62,32 @@ class _Visitor extends SimpleAstVisitor<void> {
...
-        if (field != null) {
-          if (nodeType.type == field.type) {
-            rule.reportLint(nodeType);
-          }
-        }
+    if (nodeType == null) return;
+    var paramElement = node.declaredElement;
+    if (paramElement is! FieldFormalParameterElement) return;
+
+    var field = paramElement.field;
```
Comment (scheglov) [inline lib/src/rules/type_init_formals.dart:69]: Formatting of code in this method is inconsistent with the next method.

**Review by scheglov — APPROVED** [review]

File: lib/src/rules/type_init_formals.dart (hunk 3 above)
Comment (pq) [reply lib/src/rules/type_init_formals.dart:69]: Good catch.  Fixing.  Thanks!

**Review by bwilkerson — COMMENTED** [review]
The documentation needs to be updated. I haven't been remembering to check that in other PRs, so you should go through the ones you've already landed and double check them.

**Pull request merged**